### PR TITLE
Feature 2367 filter sensitive custom headers in webscan report

### LINKED ISSUE #2367: Introduce `sensitive` tag in web scan configuration parts
## Situation
DAST Reports shall not contain sensitive/secret information like headers etc. currently the DAST report shows the given headers which can contain sensitive information.

## Wanted
It shall be possible to add something like "annotation" to the corresponding JSON parts, so its programmatically possible to identify `sensitive `information .

Also this sensivie information shall not be inside the DAST reports (OWASP ZAP)

## Solution
- TBD: Find the best location to filter the sensitive information and replace it with `**************`

### PR DESCRIPTION
**Pull request by winzj**
- closes #2367

### CODE REVIEW

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        assertEquals(data.getVulnerabilities(), maskedVulnerabilities);
+        verify(vulnerability, times(1)).getScanType();
+        verify(vulnerability, never()).getWeb();
+    }
+
+    @Test
+    void empty_web_part_results_in_unchanged_vulnerabilities() {
+        /* prepare */
+        String sechubConfigJson = TestFileReader.loadTextFile(SECHUB_WEBSCAN_CONFIG_FILE_WITH_SENSITIVE_HEADERS);
+        SecHubConfigurationModel scanConfigWithSensitiveHeaders = SecHubConfiguration.createFromJSON(sechubConfigJson);
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java:154]: Please create methods like in my examples below and use them in all of your tests (easier to read, less boilerplate code)

```java

private SechubConfigurationModel createConfigWithSensitive(){
     return createConfiguration(true);
}

private SechubConfigurationModel createConfigWithOutSensitive(){
     return createConfiguration(true);
}



private SecHubConfigurationModel createConfiguration(boolean withSensitiveHeaders){
String path = withSensitiveHeaders ? SECHUB_WEBSCAN_CONFIG_FILE_WITH_SENSITIVE_HEADERS: 
SECHUB_WEBSCAN_CONFIG_FILE_WITH_SENSITIVE_HEADERS: SECHUB_WEBSCAN_CONFIG_FILE_WITHOUT_SENSITIVE_HEADERS;

 String sechubConfigJson = TestFileReader.loadTextFile(path);
        SecHubConfigurationModel result = SecHubConfiguration.createFromJSON(sechubConfigJson);
return result
}
```

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    @Test
+    void empty_sarif_report_with_no_findings_handled_without_changing_anything() throws IOException {
+        /* prepare */
+        String sechubConfigJson = TestFileReader.loadTextFile(SECHUB_WEBSCAN_CONFIG_FILE_WITH_SENSITIVE_HEADERS);
+        SecHubConfigurationModel scanConfig = SecHubConfiguration.createFromJSON(sechubConfigJson);
+
+        String sarifReport = TestFileReader.loadTextFile(SARIF_2_1_0_OWASP_ZAP_EMPTY_REPORT_FILE);
+        SerecoMetaData data = importer.importResult(sarifReport, ScanType.WEB_SCAN);
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java:118]: Please create a method like my example below and use it inside all of your tests (makes it easier to read, understand and less boilerplate code)

```java
SerecoMetaData createTestMetaDataFromProductResult(String path){
        String productResult = TestFileReader.loadTextFile(path);
        SerecoMetaData data = importer.importResult(productResult, ScanType.WEB_SCAN);
        return data;
}
```

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+        /* execute */
+        List<SerecoVulnerability> maskedVulnerabilities = serviceToTest.maskSensitiveData(scanConfigWithSensitiveHeaders, data.getVulnerabilities());
+
+        /* test */
+        assertEquals(data.getVulnerabilities(), maskedVulnerabilities);
+        verify(vulnerability, times(1)).getScanType();
+        verify(vulnerability, never()).getWeb();
+    }
+
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java:149]: Please add tests for the illegal arguments exceptions for the null arguments for the service.

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    @Test
+    void empty_web_part_results_in_unchanged_vulnerabilities() {
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SensitiveDataMaskingServiceTest.java:151]:
```suggestion
    void null_web_part_results_in_unchanged_vulnerabilities() {
```

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java [hunk 5]
```diff
@@ -59,7 +60,7 @@ void do_import__no_importer() throws IOException {
         /* formatter:on */
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java:60]: you have added the masking service as a new dependency.
But we have no test that uses a mocked masking service...

Please add a masking service mock inside the test and
- check the masking service is called by workspace test when it comes to merging
- check merging uses the masked parts and not the origin ones (I would just check that the list from mocked service is used...)

**Review by de-jcup — CHANGES_REQUESTED** [review]
Good job, but I found some parts I would like to be changed/improved. Please look at my comments

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java [hunk 6]
```diff
@@ -59,9 +73,36 @@ void do_import__no_importer() throws IOException {
...
+    void do_import_does_call_sensitive_data_masking_service() throws IOException {
+        /* prepare */
+        SecHubMessage info = new SecHubMessage(SecHubMessageType.INFO, "info");
+        String importId = "id1";
+        String importData = TestFileReader.loadTextFile("src/test/resources/sarif/sarif_2.1.0_owasp_zap.json");
+        String productId = "PDS_WEBSCAN";
+        List<SecHubMessage> messages = List.of(info);
+
+        /* formatter:off */
+        ImportParameter importParameter = ImportParameter.builder().importData(importData).importId(importId).importProductMessages(messages)
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java:92]: You are using formatter:off here, so please indent
```suggestion
        ImportParameter importParameter =     
        
        ImportParameter.builder().
                  importData(importData).
                  importId(importId).
                  importProductMessages(messages).
                  productId(productId).
        build();
```
Or use no formatter off/on here.

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java [hunk 7]
```diff
@@ -59,9 +73,36 @@ void do_import__no_importer() throws IOException {
...
+                .productId(productId).build();
+        /* formatter:on */
+
+        List<ProductResultImporter> importers = new ArrayList<>();
+        importers.add(new SarifV1JSONImporter());
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/WorkspaceTest.java:97]: Here the sarif importer is tested also.

I would like to have here a 
```java 
ProductResultImporter fakeImporter = mock(ProductResultImporter.class);
// maybe do some stuff with fakeImporter to provide wanted behavior
importers.add(fakeImporter);

```

The advantage will be:
- no additional dependency to sarif importer/no testing of other class
- it is clear that it not ony works with the sarif importer
- you do not need to load the textfile and the other data stuff before

**Review by de-jcup — CHANGES_REQUESTED** [review]
Looks very good, except two small parts - please look at my comments.

**Review by de-jcup — APPROVED** [review]
Looks good to me :+1:
